worlds/jakanddaxter/Rules.py

```diff
@@ -0,0 +1,222 @@
+                      FireCanyonCellCount,
+                      MountainPassCellCount,
+                      LavaTubeCellCount,
+                      CitizenOrbTradeAmount,
+                      OracleOrbTradeAmount)
+from .locs import CellLocations as Cells
+from .Locations import location_table
+from .Levels import level_table
+
+
+def set_orb_trade_rule(world: JakAndDaxterWorld):
+    options = world.options
+    player = world.player
+
+    if options.enable_orbsanity == EnableOrbsanity.option_off:
+        world.can_trade = lambda state, required_orbs, required_previous_trade: (
+            can_trade_vanilla(state, player, world, required_orbs, required_previous_trade))
+    else:
+        world.can_trade = lambda state, required_orbs, required_previous_trade: (
+            can_trade_orbsanity(state, player, world, required_orbs, required_previous_trade))
+
+
+def recalculate_reachable_orbs(state: CollectionState, player: int, world: JakAndDaxterWorld) -> None:
+
+    if not state.prog_items[player]["Reachable Orbs Fresh"]:
+
+        # Recalculate every level, every time the cache is stale, because you don't know
+        # when a specific bundle of orbs in one level may unlock access to another.
+        for level in level_table:
+            state.prog_items[player][f"{level} Reachable Orbs".strip()] = (
+                count_reachable_orbs_level(state, world, level))
+
+        # Also recalculate the global count, still used even when Orbsanity is Off.
+        state.prog_items[player]["Reachable Orbs"] = count_reachable_orbs_global(state, world)
+        state.prog_items[player]["Reachable Orbs Fresh"] = True
+
+
+def count_reachable_orbs_global(state: CollectionState,
+                                world: JakAndDaxterWorld) -> int:
+
+    accessible_orbs = 0
+    for level_regions in world.level_to_regions.values():
+        for region in level_regions:
+            # Rely on short-circuiting to skip region.can_reach whenever possible.
+            if region.orb_count > 0 and region.can_reach(state):
+                accessible_orbs += region.orb_count
+    return accessible_orbs
+
+
+def count_reachable_orbs_level(state: CollectionState,
+                               world: JakAndDaxterWorld,
+                               level_name: str = "") -> int:
+
+    accessible_orbs = 0
+    for region in world.level_to_regions[level_name]:
+        # Rely on short-circuiting to skip region.can_reach whenever possible.
+        if region.orb_count > 0 and region.can_reach(state):
+            accessible_orbs += region.orb_count
+    return accessible_orbs
+
+
+def can_reach_orbs_global(state: CollectionState,
+                          player: int,
+                          world: JakAndDaxterWorld,
+                          bundle: int) -> bool:
+
+    recalculate_reachable_orbs(state, player, world)
+    return state.has("Reachable Orbs", player, world.orb_bundle_size * (bundle + 1))
```

Comment:
It would be slightly faster (and also just nicer imo) if you precalculated this `world.orb_bundle_size * (bundle + 1)` outside the function and passed that as the 4th parameter instead of `bundle`

Reply:
Same thing goes for a lot of these functions (so I won't keep mentioninig it), it would be nice if you could do a pass where you pull out any necessary "calculations" outside the function and passed the final values you need instead

Reply:
Running 50k fuzzers to make sure this refactor maintains parity, but the early results look good!

Reply:
```
Success: 9784
Failures: 0
Timeouts: 53
Ignored: 40163

Time taken:1080.22s
```